File: strings/python/valid-anagram.py
def is_anagram(s, t):
    return sorted(s) == sorted(t)


def is_anagram2(s, t):
    count = [0] * 26

    for char in s:
        count[ord(char) - ord('a')] += 1

    for char in t:
        count[ord(char) - ord('a')] -= 1

    for val in count:
        if val != 0:
            return False
    return True


print(is_anagram("anagram", "nagaram"))
print(is_anagram("batata", "attaba"))
print(is_anagram("rat", "car"))

print(is_anagram2("anagram", "nagaram"))
print(is_anagram2("batata", "attaba"))
print(is_anagram2("rat", "car"))

File: strings/java/src/ValidAnagram.java
import java.util.Arrays;

public class ValidAnagram {

    public static void main(String[] args) {

    	System.out.println(isAnagram("anagram", "nagaram"));
    	System.out.println(isAnagram("batata", "attaba"));
    	System.out.println(isAnagram("rat", "car"));

    	System.out.println(isAnagram2("anagram", "nagaram"));
    	System.out.println(isAnagram2("batata", "attaba"));
    	System.out.println(isAnagram2("rat", "car"));
    }
    
    public static boolean isAnagram(String s, String t) {
        char[] sChars = s.toCharArray();
        char[] tChars = t.toCharArray();
        
        Arrays.sort(sChars);
        Arrays.sort(tChars);
        
        return Arrays.equals(sChars, tChars);
    }
    
    public static boolean isAnagram2(String s, String t) {

        int[] count = new int[26];
        
        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
        }
        
        for (int i = 0; i < t.length(); i++) {
            count[t.charAt(i) - 'a']--;
        }
        
        for (int c : count) {
            if (c != 0) {
                return false;
            }
        }
        
        return true;
    }
}
